### PR TITLE
integration for web-poet's support on additional requests and Meta

### LINKED ISSUE #65: `callback_for` does not produce a coroutine.
Looks like `callback_for` doesn't produce parse method that is a coroutine. So it's not possible to use `callback_for` in case `to_item` method is a coroutine (declared with async/await syntax).

Example:
```python
import asyncio
import inspect

from scrapy_poet import callback_for
from web_poet.pages import ItemWebPage


class SimplePage(ItemWebPage):
    def to_item(self):
        return {}

cb = callback_for(SimplePage)
print(f"Is callback a coroutine: {inspect.iscoroutinefunction(cb)}")
print(f"Is to_item a coroutine: {inspect.iscoroutinefunction(SimplePage.to_item)}")

class SimplePageAIO(ItemWebPage):
    async def to_item(self):
        await asyncio.sleep(0.1)
        return {}

cb_aio = callback_for(SimplePageAIO)
print(f"Is callback a coroutine: {inspect.iscoroutinefunction(cb_aio)}")
print(f"Is to_item a coroutine: {inspect.iscoroutinefunction(SimplePageAIO.to_item)}")
```
This produces:
```
Is callback a coroutine: False
Is to_item a coroutine: False 
Is callback a coroutine: False
Is to_item a coroutine: True
```
But it should be like:
```
Is callback a coroutine: False
Is to_item a coroutine: False 
Is callback a coroutine: True
Is to_item a coroutine: True
```
Tested on `dev-1` branch.

### PR DESCRIPTION
**Pull request by BurnzZ**
Requires the following to be merged in `web-poet`:
- https://github.com/scrapinghub/web-poet/pull/22
- https://github.com/scrapinghub/web-poet/pull/33

This is built on top of 
- https://github.com/scrapinghub/scrapy-poet/pull/67

### Progress

- [x] Testing across a wide-variety of scenarios/cases
     - contextvars and Twisted
     - scrapy-poet and scrapy integration
- [x] Docs
    - [x] update doc links using intersphinx-mapping when `web-poet` is released.
- [x] update error handling after https://github.com/scrapinghub/web-poet/pull/38 is merged
    - [x] ensure that Scrapy doesn't redirect on `HEAD` requests ([reference](https://github.com/scrapinghub/web-poet/pull/38#discussion_r867297335))
- [x] Tests
    - [x] Add test to verify that additional requests work both with an asyncio and a non-asyncio Twisted reactor.
- [x] Changelog
- [x] Use dont_filter for additional requests
- [x] Document that spider middlewares are not followed for additional requests, while downloader middlewares are.
- Once there is a web-poet release with the new page input classes, like `HttpResponse`:
    - [x] Update minimum dependencies `web-poet>=0.2.0` and `scrapy>=2.6.0` in **setup.py** and **tox.ini**

Fixes #65.

### CODE REVIEW

File: tox.ini
Comment (BurnzZ) [inline tox.ini:10]: **TODO**: Reminder to revert this change before merging.

File: scrapy_poet/middleware.py
Comment (kmike) [inline scrapy_poet/middleware.py:61]: How does it work with multiple spiders in the same process? It seems this might be a problem. I'm not sure about the solution yet.

Comment (BurnzZ) [reply scrapy_poet/middleware.py:61]: I've tested this out but couldn't find any problem as running multiple spiders in the same process works okay.

@kmike I've tagged you on a separate PR which shows the simple test I've made. Perhaps there's some problem or behavior change that I've missed out.

Comment (kmike) [reply scrapy_poet/middleware.py:61]: answered there, thanks @BurnzZ!

Comment (BurnzZ) [reply scrapy_poet/middleware.py]: Hi @kmike , cheers again for raising. 

After some deep investigation on this matter, it would seem that **Twisted** only has a limited support for **ContextVars** at the moment. It currently breaks some of our expectations in using **ContextVars** inside Scrapy. _(see section below for full explanation)_

With that, I've updated the approach inside `scrapy-poet` to use **Dependency Injection** instead. This creates a consistent non-leaking value across different multiple CrawlerProcesses. Here's the commit for this: https://github.com/scrapinghub/scrapy-poet/pull/62/commits/e3b96dce17b199ebe3804aa68f20a1e6ee3b96d0

### Investigation Notes
---------------------

It would seem that having this order of multiple spiders in a single process:

```python
process = CrawlerProcess(settings)
process.crawl(SpiderAAA)
process.crawl(SpiderBBB)
process.crawl(SpiderCCC)
process.start()
```

would result in:

- ✅ **SpiderAAA** using the downloader from **SpiderAAA** itself
- ❌ **SpiderBBB** using the downloader from **SpiderAAA**
- ❌ **SpiderCCC** using the downloader from **SpiderAAA**

In other words, when the first spider sets the downloader, it would be used throughout in the succeeding spiders, regardless if the succeeding spiders attempt to update the ContextVars directly. I've tried it with more spiders and the behavior is simply the same.

We can override the `ContextVar` value inside the spider though. However, after some testing, it would seem that Twisted leaks the context outside of each crawler process. So if we update the `ContextVar` value inside **SpiderBBB**, it would leak as:

- ✅ **SpiderAAA** using the downloader from **SpiderAAA** itself
- ✅ **SpiderBBB** using the downloader from **SpiderBBB**
- ❌ **SpiderCCC** using the downloader from **SpiderBBB**


Interestingly, the behavior above only happens when `twisted>=21.2.0` since it introduced this change:
> twisted.internet.defer.inlineCallbacks and ensureDeferred will now associate a contextvars.Context with the coroutines they run, meaning that ContextVar objects will maintain their value within the same coroutine, similarly to asyncio Tasks. This functionality requires Python 3.7+, or the contextvars PyPI backport to be installed for Python 3.5-3.6. (#9719, #9826)

Here's the code comment in Twisted that tests the ContextVars in InlineCallbacks ([reference](https://github.com/twisted/twisted/pull/1192/files#diff-d51be359d7730f86734b450ecbf6dc8f54090e84e7efcf5616399fb5cffc96a7R3114-R3118)).

> When an inlineCallbacks function is called, the context is taken from
> when it was first called. When it resumes, the same context is applied.

A twisted version below `21.2.0` would result in the the following as expected ([Reference](https://github.com/twisted/twisted/releases/tag/twisted-21.2.0)):

- ✅ **SpiderAAA** using the downloader from **SpiderAAA** itself
- ✅ **SpiderBBB** using the downloader from **SpiderBBB**
- ✅ **SpiderCCC** using the downloader from **SpiderCCC**

However, the expected correct behavior above only happens during normal spider operation. When the time comes where an Additional Request is to be executed in the downloader, the ContextVar raises a `LookupError` since the context is not preserved. ❌

After digging further, it would seem that Twisted doesn't support all of the ContextVar features completely:

- [Support contextvars in Deferred](https://twistedmatrix.com/trac/ticket/9807)
- [Support contextvars in DelayedCall](https://twistedmatrix.com/trac/ticket/9824)
- [Contextvars support does not support `.reset`](https://twistedmatrix.com/trac/ticket/10301)

File: CHANGELOG.rst
Comment (BurnzZ) [inline CHANGELOG.rst:8]: Reminder to confirm this before merging since the anticipated `web_poet` version release might change.

File: docs/intro/advanced-tutorial.rst
Comment (kmike) [inline docs/intro/advanced-tutorial.rst:134]: I think the fact that @BurnzZ you felt a need to call a key for web_poet Meta "``po_args``" means we should think about web_poet.Meta name again. Ideally, names should match.

Comment (BurnzZ) [reply docs/intro/advanced-tutorial.rst:135]: Hmmm, yes, a good point.

One option is to rename the `Meta` class as `PoArgs` to semantically match `request.meta["po_args"]`.

Maybe other names could be `PageArgs`, `PageParams`, and `PoParams`.

Comment (Gallaecio) [reply docs/intro/advanced-tutorial.rst:135]: This is something to change in web-poet itself, if we change it at all, right?

We could also use `request.meta["po_meta"]`. I don’t see a big issue with it, I think the repeated `meta` is acceptable in this scenario. We could use `request.meta["po_meta"]`, but we would need to be sure that we will never need additional page object data handled through Scrapy request metadata.

Comment (BurnzZ) [reply docs/intro/advanced-tutorial.rst:135]: > This is something to change in web-poet itself, if we change it at all, right?

Yes, [this one](https://github.com/scrapinghub/scrapy-poet/pull/62/files#diff-a38b072945ddeae120d2462383b70ddc0f40967b53c849c49d6a3cbdf1917f53R217-R225) in particular, which is at its core.

> but we would need to be sure that we will never need additional page object data handled through Scrapy request metadata

I think it's safe to say that it will be needed, for cases [like this](https://github.com/scrapinghub/scrapy-poet/pull/62/files#diff-261312b8086ac4bda934c1858d1479ee18c82ba2c22615f1f0ccc798a0dc30b4R162).

Comment (Gallaecio) [reply docs/intro/advanced-tutorial.rst:135]: > We could use `request.meta["po_meta"]`, but we would need to be sure that we will never need additional page object data handled through Scrapy request metadata.

I made a typo here :facepalm: . I meant `"po"`. We could use `"po"` if we are sure po-related stuff in `request.meta` will always be limited to `po.Meta` data. Which I would not bet on, even if it seems so now.

File: scrapy_poet/backend.py
Comment (Gallaecio) [inline scrapy_poet/backend.py:41]: I am not sure how to best approach unexpected exceptions. Here I am assuming they are requests errors (twisted exceptions, openssl ones, etc.), but they could as well come from downloader middlewares.

By the way, I thought I would need to handle https://github.com/scrapy/scrapy/blob/master/scrapy/spidermiddlewares/httperror.py#L13 here and map it to `HttpResponseError`, but it turns out it is a spider middleware, not a downloader middleware, so it is not executed for additional requests, and the default status code handling by `HttpClient` works.

Comment (BurnzZ) [reply scrapy_poet/backend.py:44]: Yeah, this is indeed quite tricky. For example, the retry middleware handles a [bunch of different request errors](https://github.com/scrapy/scrapy/blob/master/scrapy/downloadermiddlewares/retry.py#L131-L134) from twisted but [returns None](https://github.com/scrapy/scrapy/blob/master/scrapy/downloadermiddlewares/retry.py#L124) when max retries are reached.

I wonder if it would be safe to assume that if there was no response (e.g. max retries reached) then we can raise `HttpRequestError`.

File: scrapy_poet/backend.py
Comment (kmike) [inline scrapy_poet/backend.py:43]: I wonder if we should use `raise from` when re-raising exceptions, so that the original Scrapy's exception is available as `ex.__cause__`.

I'm a bit confused about it overall, because `ex.__context__` is also set explicitly, but it seems the semantics are a bit different.

**Comment by BurnzZ** [comment]
Hi @Gallaecio , thanks for wrapping this up and bringing it to the finish line! 🙌 LGTM! 🚀

File: scrapy_poet/backend.py
Comment (kmike) [inline scrapy_poet/backend.py:27]: I think it'd be good to create an explicit function to convert from web_poet.HttpRequest to Scrapy's Request, and add tests for it - check that headers work, etc. 
``attr.asdict(request)`` looks a bit risky; it'd also break after we introduce URL classes.

File: scrapy_poet/backend.py
Comment (kmike) [inline scrapy_poet/backend.py:31]: I wonder how can users act on this suggestion

File: scrapy_poet/backend.py
Comment (kmike) [inline scrapy_poet/backend.py:41]: Could you please add a comment, explaining when this error can happen?

File: scrapy_poet/utils.py
Comment (kmike) [inline scrapy_poet/utils.py:25]: This is probably a rare edge case, but it seems the conversion won't work properly for TextResponses where encoding is set explicitly, in `__init__` method. We're not passing this encoding, which means HttpResponse would try to guess the encoding again, and the result could be different.

File: scrapy_poet/utils.py
Comment (kmike) [inline scrapy_poet/utils.py:21]:
```suggestion
def scrapy_response_to_http_response(response: Response) -> HttpResponse:
```

File: scrapy_poet/backend.py
Comment (Gallaecio) [reply scrapy_poet/backend.py:31]: Should we expose a way (e.g. a setting) for users to provide their custom implementation of web poet request to Scrapy request conversion?

Or should we maybe reconsider how we handle this? For example, ignore extra fields, possibly with a warning.

Comment (kmike) [reply scrapy_poet/backend.py:31]: I'm not sure when could the user get this message tbh. web_poet.HttpRequest is a well-defined class, scrapy.Request is a well-defined class, why would there be more than one way to convert between those?

File: scrapy_poet/page_input_providers.py
Comment (kmike) [inline scrapy_poet/page_input_providers.py:225]: Would this work? I wonder if some of the keys may not be valid argument names.

```suggestion
        return [Meta(request.meta.get("po_meta", {}))]
```

File: tests/test_backend.py
Comment (kmike) [inline tests/test_backend.py:151]: I wonder how hard would it be to have integration-style tests for such features, similar to https://github.com/scrapy-plugins/scrapy-splash/blob/f5273b3a0ef15a9f6809d305a1dbe8f3efd12c5f/tests/test_integration.py#L113. My worry is that Scrapy has many components, which interact in various ways, so it's possible we miss something if we rely on mock-style tests. What do you think, would it worth it?

File: scrapy_poet/page_input_providers.py
Comment (Gallaecio) [reply scrapy_poet/page_input_providers.py:225]: TIL

```python
>>> dict(1=2)
  File "<stdin>", line 1
SyntaxError: keyword can't be an expression
>>> dict(**{"1": "2"})
{'1': '2'}
```

So I guess we could leave the `**` there, but it is probably better to remove it unless there is a good reason to keep it.

File: scrapy_poet/backend.py
Comment (Gallaecio) [reply scrapy_poet/backend.py]: I guess we can assume there will be no HttpRequest subclasses requiring special conversion logic. We can always revisit in the future if someone comes up with a valid use case.

File: docs/intro/advanced-tutorial.rst
Comment (kmike) [reply docs/intro/advanced-tutorial.rst]: hey! Before we close it off, let's do a final check on the argument/dependency name. It affects both web-poet and scrapy-poet.

I think two main options are:

1. keep web_poet.Meta. This is similar to Scrapy's request.meta, so it can be familiar for developers who worked with Scrapy (?).
2. rename it to something like web_poet.Arguments, web_poet.PageParams, etc. 

After giving it some time to settle, I think I slighly (+0.5) prefer the second option, because of these reasons:

1. In most examples we provide, this argument ("po_meta") is used to control Page behavior. It's some additional per-request parameters, arguments we pass to a page object, which is specific to this page object. So, something like "arguments" or "params" make it more explicit.
2. In general, "meta" or "metadata" is some information about other information - about request in our case. Common examples of metadata are "date created", "author", etc. It's commonly used to aggregate information later, or to provide some additional information about a data point. I'm not sure how common is it to use "meta" or "metadata" to describe arguments which control the behavior; we don't call function parameters "metadata". In Scrapy this happens, but even there meta is used to provide after-the-fact information about the request as well (e.g. download_latency). Also, some uses of Scrapy's meta are deprecated in favor of cb_kwargs.
3. Naming it "meta" could lead to confusion with Scrapy's request.meta. First, user might expect that by depending on web_poet.Meta you'd get access to request.meta, which is not the case. Second, request.meta is mutable - for components like middlewares it can be used both as an input (to control the behavior), and as an output (to provide some information about what happened), while web_poet.Meta is not intended to be mutable, it's only there to be read. So, even for Scrapy users not having "Meta" in the name could allow to reduce the possible confusion.

----

Unrealted to the above, I think we shouldn't use "po" shortcut. We moved away from it (e.g. scrapy-poet used to be called scrapy-po), because it doesn't always sound great in English, some people were saying it like "poo" :)

Comment (BurnzZ) [reply docs/intro/advanced-tutorial.rst]: You've raised great points @kmike.

I'm inclined to rename `web_poet.Meta` into `web_poet.PageParams`. From the naming semantics, it's quite clear that it contains parameters for a page. In addition, perhaps we could also rename `request.meta["po_args"]` into `request.meta["page_params"]` for symmetry.

File: tests/test_backend.py
Comment (Gallaecio) [inline tests/test_backend.py:371]: Is it OK to leave this out of the scope of this pull request, or should I look into addressing this here?

**Comment by Gallaecio** [comment]
Today I (Re-)Learn duplicate filtering happens on the scheduler, so we do not need to use `dont_filter=True` for additional requests because they already bypass the scheduler. Integration tests have been worth it for sure.

File: docs/intro/advanced-tutorial.rst
Comment (kmike) [inline docs/intro/advanced-tutorial.rst:146]:
```suggestion
    class ProductSpider(scrapy.Spider):
```

File: scrapy_poet/backend.py
Comment (kmike) [inline scrapy_poet/backend.py:19]: Would it make sense to call the argument as something like `download_func`, not `backend`?

File: tests/test_backend.py
Comment (kmike) [reply tests/test_backend.py:351]: I think that's ok.

File: tests/test_backend.py
Comment (kmike) [inline tests/test_backend.py:441]: it seems this test doesn't check that 2 requests in to_item are made, it could pass if the second (or both) of the requests were no-op, without raising an exception

File: docs/intro/advanced-tutorial.rst
Comment (Gallaecio) [inline docs/intro/advanced-tutorial.rst:168]: Shouldn’t this be more like:

```python
                    meta={"page_params": {ProductPage: {"enable_extracting_all_images": True}}}
```

So that we can have page-specific params?

```python
                    meta={
                        "page_params": {
                            ProductPage: {"enable_extracting_all_images": True},
                            BrowserHtml: {"geo": "FR"}
                        }
                    }
                )

        async def parse(self, response, html: browserHtml, page: ProductPage):
```

Comment (BurnzZ) [reply docs/intro/advanced-tutorial.rst:168]: Yes, this interface makes sense to me. 👍

Comment (Gallaecio) [reply docs/intro/advanced-tutorial.rst:168]: After an off-GitHub discussion with @BurnzZ and @kmike, we decided against this because at the moment there is no use case where the same param should have a different value for different page objects for the same request. If that ever changes, maybe we can revisit.

**Review by kmike — APPROVED** [review]
👍 +1 to merge it after making sure it works with the latest web-poet (https://github.com/scrapinghub/web-poet/pull/42 is merged).

**Comment by Gallaecio** [comment]
Closing and reopening to trigger tests with the latest web-poet master.

As for merging, we still need to wait for a web-poet release (0.2?), then update requirements here, re-run tests, and then merge.

**Comment by kmike** [comment]
@Gallaecio could you please check https://github.com/scrapinghub/web-poet/pull/51 before making scrapy-poet release with these changes?

**Comment by Gallaecio** [comment]
https://github.com/scrapinghub/scrapy-poet/pull/62/commits/3398dc77f96bfdbffaa104612547252a730f7a51 is not as trivial as the commit message may suggest:
- Added the right minimum versions of dependencies to `setup.py` (lowest versions for which tests pass)
- Had tox test those minimum versions (I have just realized that I need to update GitHub Actions accordingly in a follow-up commit)
- Made Scrapy a dependency in line with https://github.com/scrapy-plugins/scrapy-zyte-api/pull/29
- Removed tldextract as a dependency, not needed since https://github.com/scrapinghub/scrapy-poet/pull/56 (although it is still available, as it is a dependency of Scrapy)
- mypy now runs with the minimum version of Python and dependencies.

File: tox.ini
Comment (Gallaecio) [reply tox.ini:42]: My concern was that we could be using typing hints incompatible with the latest Python, but come to think of it, thouse should cause SyntaxError on the regular Python 3.7 tests, so maybe that is fine.

File: setup.py
Comment (BurnzZ) [inline setup.py:20]: As per https://github.com/scrapinghub/web-poet/pull/51#issuecomment-1154937042, we should probably make this into `0.3.0`, as well as updating the CHANGELOG and tox.ini.

**Review by BurnzZ — COMMENTED** [review]
Got one minor change left and I believe we're good to go. 🚀

File: setup.py
Comment (Gallaecio) [reply setup.py:20]: Yes, I went through the same path. It was when I went to apply the contextvar name change to the patch when I realized we do not need 0.3.0 here.

**Comment by kmike** [comment]
Thanks @BurnzZ and @Gallaecio, that's a big milestone!

**Pull request merged**
